src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SslOverTdsStream.cs

```diff
@@ -89,66 +89,51 @@ public override Task<int> ReadAsync(byte[] buffer, int offset, int count, Cancel
         /// </summary>
         private async Task<int> ReadInternal(byte[] buffer, int offset, int count, CancellationToken token, bool async)
         {
+            int readBytes = 0;
+            byte[] packetData = null;
+            byte[] readTarget = buffer;
+            int readOffset = offset;
             if (_encapsulate)
             {
-                return await ReadInternalEncapsulate(buffer, offset, count, token, async).ConfigureAwait(false);
-            }
-            else if (async)
-            {
-                return await ReadInternalAsync(buffer, offset, count, token).ConfigureAwait(false);
-            }
-            else
-            {
-                return ReadInternalSync(buffer, offset, count);
-            }
-        }
+                packetData = ArrayPool<byte>.Shared.Rent(count < TdsEnums.HEADER_LEN ? TdsEnums.HEADER_LEN : count);
+                readTarget = packetData;
+                readOffset = 0;
+                if (_packetBytes == 0)
+                {
+                    // Account for split packets
+                    while (readBytes < TdsEnums.HEADER_LEN)
+                    {
+                        readBytes += async ?
+                            await _stream.ReadAsync(packetData, readBytes, TdsEnums.HEADER_LEN - readBytes, token).ConfigureAwait(false) :
+                            _stream.Read(packetData, readBytes, TdsEnums.HEADER_LEN - readBytes);
+                    }
 
-        private async Task<int> ReadInternalEncapsulate(byte[] buffer, int offset, int count, CancellationToken token, bool async)
-        {
-            int readBytes = 0;
-            byte[] packetData = ArrayPool<byte>.Shared.Rent(count < TdsEnums.HEADER_LEN ? TdsEnums.HEADER_LEN : count);
+                    _packetBytes = (packetData[TdsEnums.HEADER_LEN_FIELD_OFFSET] << 8) | packetData[TdsEnums.HEADER_LEN_FIELD_OFFSET + 1];
+                    _packetBytes -= TdsEnums.HEADER_LEN;
+                }
 
-            if (_packetBytes == 0)
-            {
-                // Account for split packets
-                while (readBytes < TdsEnums.HEADER_LEN)
+                if (count > _packetBytes)
                 {
-                    readBytes += (async ?
-                        await ReadInternalAsync(packetData, readBytes, TdsEnums.HEADER_LEN - readBytes, token).ConfigureAwait(false) :
-                        ReadInternalSync(packetData, readBytes, TdsEnums.HEADER_LEN - readBytes)
-                   );
+                    count = _packetBytes;
                 }
-
-                _packetBytes = (packetData[TdsEnums.HEADER_LEN_FIELD_OFFSET] << 8) | packetData[TdsEnums.HEADER_LEN_FIELD_OFFSET + 1];
-                _packetBytes -= TdsEnums.HEADER_LEN;
             }
 
-            if (count > _packetBytes)
+            readBytes = async ?
+                await _stream.ReadAsync(readTarget, readOffset, count, token).ConfigureAwait(false) :
```

Comment:
I'm comparing these changes with the code before the problem PR and 122 and 123 essentially go from:
Read(readTarget, 0, count)
to
Read(readTarget, readOffset, count)
where readOffset is 0 if _encapsulate == true but readOffset == offset if _encapsulate == false. So I'm wondering if the behavior change when _encapsulate == false is intentional. Offset seems to be accounted for on line 131 which remains the same as before the problem PR.

I'll admit I'm not too familiar with this part of the code but I'm trying to take a close look given the circumstances.

Reply:
The behaviour doesn't change if `_encapsulate` is false because the read* variables are initialized to the buffer and offset. 

If _encapsulate==false it'll be setup to read directly into the array location from the parameters
if _encapsulate==true the read* variables are set to the rented location. At very end if there is a rented buffer it copies back from the rented location to the parameters location.